test/1_stdlib/UnfoldSequence.swift

```diff
@@ -0,0 +1,41 @@
+// RUN: %target-run-simple-swift
+// REQUIRES: executable_test
+
+import StdlibUnittest
+
+let UnfoldTests = TestSuite("UnfoldSequence")
+
+UnfoldTests.test("sequence(state:next:)") {
+  // FIXME: The full type signatures on these closures should not be
+  // necessary, but at the moment the compiler gives very confusing errors if
+  // we don't have them.
+  let s0 = sequence(state: 0, next: { (val: inout Int) -> Int? in val += 1; return val < 10 ? val : nil })
+  expectEqualSequence(1..<10, s0)
+
+  let s1 = sequence(state: (), next: { (_: inout ()) -> Int? in 1 })
+  expectEqualSequence([1, 1, 1], s1.prefix(3))
+
+  let s2 = sequence(state: (1..<6).makeIterator(), next: { (iter: inout CountableRange<Int>.Iterator) in iter.next() })
+  expectEqualSequence(1..<6, s2)
+
+  var calls = 0
+  var s3 = sequence(state: 0, next: { (val: inout Int) -> Int? in calls += 1; val += 1; return val })
+  for i in 1..<6 {
+    expectEqual(i, s3.next())
+    expectEqual(i, calls)
+  }
+}
+
+UnfoldTests.test("sequence(first:next:)") {
+  let s0 = sequence(first: 1, next: { $0 < 50 ? $0 * 2 : nil })
+  expectEqualSequence([1, 2, 4, 8, 16, 32, 64], s0)
+
+  var calls = 0
+  var s1 = sequence(first: 0, next: { calls += 1; return $0 + 1 })
+  for i in 0..<5 {
+    expectEqual(i, s1.next())
+    expectEqual(i, calls)
+  }
+}
+
```

Comment:
@gribozavr It looks like `checkSequence()` also checks to make sure that invoking `next()` 10 times after it returned `nil` continues to return `nil` for those 10 times. This is checking something that isn't currently a requirement of `Iterator` according to the doc comment (I know there was talk about changing this some time ago but I'm not sure what the outcome of that was). Unfortunately, matching this behavior with `sequence(state:next:)` requires an extra branch on every call to `next()`, so I'd rather not implement this behavior without a good reason. What would you prefer me to do, take the cost of the branch so I can use `checkSequence()`, or skip the call to `checkSequence()` in this case? AFAICT the only thing `checkSequence()` is really checking for besides this is that the actual count of the sequence is >= the `underestimatedCount` (it also has a check with `_preprocessingPass` but `UnfoldSequence` is destructively-iterated so that check will be skipped).